examples/tutorials/alpha-zero-tic-tac-toe/DeterminedShim.py

```diff
@@ -0,0 +1,144 @@
+        if ctx is not None:
+            ctx.__exit__(None, None, None)
+            setattr(self, attr, None)
+
+    def _latest_checkpoint(self):
+        if self.info is None:
+            return None
+        if self.info.latest_checkpoint is not None:
+            return self.info.latest_checkpoint
+        if "latest_checkpoint" in self.info.user_data:
+            return self.info.user_data["latest_checkpoint"]
+        return None
+
+    def _stage_restore(self, required=True):
+        latest_checkpoint = self._latest_checkpoint()
+        if latest_checkpoint is None:
+            return
+        try:
+            self.restore = self.context.checkpoint.restore_path(latest_checkpoint)
+            self.restore_path = self.restore.__enter__()
+            self.restore_meta = self.context.checkpoint.get_metadata(latest_checkpoint)
+            if "steps_completed" in self.restore_meta:
+                self.steps = self.restore_meta["steps_completed"]
+            log.info("checkpoint.get_metadata(%s) -> %r", latest_checkpoint, self.restore_meta)
+        except Exception as err:
+            log.error("unable to load latest_checkpoint: %r", err)
+            if required:
+                raise err
+
+    def step(self, metadata=None):
+        self.steps += 1
+        if self.info is None:
+            return
+        self._release("storage")
+        if self.context.preempt and self.context.preempt.should_preempt():
+            log.warn("task early exiting on Determined preempt")
+            sys.exit(0)
+        if metadata is None:
+            metadata = {}
+        if not "steps_completed" in metadata:
+            metadata["steps_completed"] = self.steps
+        self.storage = self.context.checkpoint.store_path(metadata)
+        (self.store_path, _storage_id) = self.storage.__enter__()
+        log.info("step: checkpoint.store_path(%r) -> %s", metadata, self.store_path)
+
+    def save_path(self, folder):
+        if self.store_path is None:
+            return pathlib.Path(folder)
+        return self.store_path.joinpath(folder)
+
+    def load_path(self, folder):
+        if self.restore_path is None:
+            return (pathlib.Path(folder), {})
+        return (self.restore_path.joinpath(folder), self.restore_meta)
+
+    def training_metrics(self, history: Dict[str, Any]):
+        if self.context.train is None:
+            return
+        batch_metrics = []
+        for key, items in history.items():
+            for i, val in enumerate(items):
+                if i >= len(batch_metrics):
+                    batch_metrics.insert(i, {})
+                batch_metrics[i][key] = val
+        self.context.train.report_training_metrics(self.steps, batch_metrics[-1], batch_metrics)
+
+    def validation_metrics(self, metrics):
+        if self.context.train is None:
+            return
+        self.context.train.report_validation_metrics(self.steps, metrics)
+
+    # help: need stable way to get max_length from core api
```

Comment:
I think using the experiment configuration is fine. Or is this an older comment from before you used the config?

Reply:
The problem is that using `trial._config` should be an internal only API. Ideally the searcher will pick a length, up to max_length, to be used in the trial. In that way it might be nice to have a function like `trial.length()`. 

I spoke with @rb-determined-ai briefly about this yesterday, not sure of our conclusion but pinging here as it feels like there is room for some improvement.